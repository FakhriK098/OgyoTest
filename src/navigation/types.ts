import { NavigatorScreenParams } from '@react-navigation/native';

export type RootStackParamList = {
  MainTabs: NavigatorScreenParams<MainTabParamList>;
  Detail: { id: string; title: string; data?: any };
};

export type HomeStackParamList = {
  Detail: { id: string };
};

export type MainTabParamList = {
  Home: undefined;
  User: undefined;
};
