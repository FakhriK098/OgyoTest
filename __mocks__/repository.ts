import { IRepository } from 'src/types/repository';

export const dummyRepositories: IRepository[] = [
  {
    id: 1,
    node_id: 'MDEwOlJlcG9zaXRvcnkx',
    name: 'hello-world',
    full_name: 'octocat/hello-world',
    private: false,
    owner: {
      login: 'octocat',
      id: 1,
      node_id: 'MDQ6VXNlcjE=',
      avatar_url: 'https://avatars.githubusercontent.com/u/1?v=4',
      gravatar_id: '',
      url: 'https://api.github.com/users/octocat',
      html_url: 'https://github.com/octocat',
      followers_url: 'https://api.github.com/users/octocat/followers',
      following_url:
        'https://api.github.com/users/octocat/following{/other_user}',
      gists_url: 'https://api.github.com/users/octocat/gists{/gist_id}',
      starred_url:
        'https://api.github.com/users/octocat/starred{/owner}{/repo}',
      subscriptions_url: 'https://api.github.com/users/octocat/subscriptions',
      organizations_url: 'https://api.github.com/users/octocat/orgs',
      repos_url: 'https://api.github.com/users/octocat/repos',
      events_url: 'https://api.github.com/users/octocat/events{/privacy}',
      received_events_url:
        'https://api.github.com/users/octocat/received_events',
      type: 'User',
      user_view_type: 'public',
      site_admin: false,
    },
    html_url: 'https://github.com/octocat/hello-world',
    description: 'A simple hello world repository',
  },
  {
    id: 2,
    node_id: 'MDEwOlJlcG9zaXRvcnky',
    name: 'awesome-project',
    full_name: 'johndoe/awesome-project',
    private: false,
    owner: {
      login: 'johndoe',
      id: 2,
      node_id: 'MDQ6VXNlcjI=',
      avatar_url: 'https://avatars.githubusercontent.com/u/2?v=4',
      gravatar_id: '',
      url: 'https://api.github.com/users/johndoe',
      html_url: 'https://github.com/johndoe',
      followers_url: 'https://api.github.com/users/johndoe/followers',
      following_url:
        'https://api.github.com/users/johndoe/following{/other_user}',
      gists_url: 'https://api.github.com/users/johndoe/gists{/gist_id}',
      starred_url:
        'https://api.github.com/users/johndoe/starred{/owner}{/repo}',
      subscriptions_url: 'https://api.github.com/users/johndoe/subscriptions',
      organizations_url: 'https://api.github.com/users/johndoe/orgs',
      repos_url: 'https://api.github.com/users/johndoe/repos',
      events_url: 'https://api.github.com/users/johndoe/events{/privacy}',
      received_events_url:
        'https://api.github.com/users/johndoe/received_events',
      type: 'User',
      user_view_type: 'public',
      site_admin: false,
    },
    html_url: 'https://github.com/johndoe/awesome-project',
    description: 'An awesome project to test API responses',
  },
  {
    id: 3,
    node_id: 'MDEwOlJlcG9zaXRvcnkz',
    name: 'cool-library',
    full_name: 'janedoe/cool-library',
    private: true,
    owner: {
      login: 'janedoe',
      id: 3,
      node_id: 'MDQ6VXNlcjM=',
      avatar_url: 'https://avatars.githubusercontent.com/u/3?v=4',
      gravatar_id: '',
      url: 'https://api.github.com/users/janedoe',
      html_url: 'https://github.com/janedoe',
      followers_url: 'https://api.github.com/users/janedoe/followers',
      following_url:
        'https://api.github.com/users/janedoe/following{/other_user}',
      gists_url: 'https://api.github.com/users/janedoe/gists{/gist_id}',
      starred_url:
        'https://api.github.com/users/janedoe/starred{/owner}{/repo}',
      subscriptions_url: 'https://api.github.com/users/janedoe/subscriptions',
      organizations_url: 'https://api.github.com/users/janedoe/orgs',
      repos_url: 'https://api.github.com/users/janedoe/repos',
      events_url: 'https://api.github.com/users/janedoe/events{/privacy}',
      received_events_url:
        'https://api.github.com/users/janedoe/received_events',
      type: 'User',
      user_view_type: 'public',
      site_admin: false,
    },
    html_url: 'https://github.com/janedoe/cool-library',
    description: 'A private library for internal usage',
  },
];
