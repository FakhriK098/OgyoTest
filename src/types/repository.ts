export interface IOwner {
  login: string;
  id: number;
  node_id: string;
  avatar_url: string;
  gravatar_id: string;
  url: string;
  html_url: string;
  followers_url: string;
  following_url: string;
  gists_url: string;
  starred_url: string;
  subscriptions_url: string;
  organizations_url: string;
  repos_url: string;
  events_url: string;
  received_events_url: string;
  type: string;
  user_view_type: string;
  site_admin: boolean;
}

export interface IRepository {
  id: number;
  node_id: string;
  name: string;
  full_name: string;
  private: boolean;
  owner: IOwner;
  html_url: string;
  description: string;
}

export interface IFetchRepositoriesPayload {
  since?: number;
}

export interface ISearchRepositoriesPayload {
  query: string;
}

export interface ILinkHeader {
  next?: string;
  prev?: string;
  first?: string;
  last?: string;
}
