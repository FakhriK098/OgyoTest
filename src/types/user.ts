export interface IUser {
  id: number;
  node_id: string;
  name: string;
  full_name: string;
  private: boolean;
  owner: IUserOwner;
  html_url: string;
  description: string;
  size: number;
  stargazers_count: number;
  watchers_count: number;
  language: string;
  forks_count: number;
  open_issues_count: number;
  license: ILicense;
  topics: string[];
  network_count: number;
  subscribers_count: number;
}

export interface ILicense {
  key: string;
  name: string;
  spdx_id: string;
  url: string;
  node_id: string;
}

export interface IUserOwner {
  login: string;
  id: number;
  node_id: string;
  avatar_url: string;
  gravatar_id: string;
  url: string;
  html_url: string;
  followers_url: string;
  following_url: string;
  gists_url: string;
  starred_url: string;
  subscriptions_url: string;
  organizations_url: string;
  repos_url: string;
  events_url: string;
  received_events_url: string;
  type: string;
  user_view_type: string;
  site_admin: boolean;
}

export interface IFetchUserPayload {
  fullName?: string;
}

export interface ILabelProps {
  title: string;
  value?: string;
}
