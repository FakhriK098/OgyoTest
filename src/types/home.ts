import { IRepository } from './repository';

export interface IFloatingActionButtonProps {
  onPress: () => void;
}

export interface ISearchBarProps {
  onSearch: (query: string) => void;
  onSelectSuggestion: (repository: IRepository) => void;
  onClear: () => void;
  suggestions: IRepository[];
  loading: boolean;
  placeholder?: string;
  onSubmitSearch: (query: string) => void;
}

export type TSortOrder = 'ASC' | 'DESC';

export interface ISortFilterModalProps {
  visible: boolean;
  onClose: () => void;
  onSort: (order: TSortOrder) => void;
  currentOrder: TSortOrder;
}
