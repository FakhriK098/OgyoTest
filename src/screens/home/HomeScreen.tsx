import React, { useEffect, useState, useMemo } from 'react';
import {
  View,
  Text,
  StyleSheet,
  SafeAreaView,
  FlatList,
  ActivityIndicator,
  RefreshControl,
} from 'react-native';
import { colors } from '../../themes/colors';
import { useAppDispatch, useAppSelector } from '../../hooks/redux';
import {
  fetchRepositoriesRequest,
  fetchMoreRepositoriesRequest,
  fetchSearchRepositoriesRequest,
  clearSearchResults,
} from '../../store/slices/repositorySagaSlice';
import { IRepository } from '../../types/repository';
import CardRepository from './components/CardRepository';
import ItemSeparator from '../../components/ItemSeparator';
import FloatingActionButton from './components/FloatingActionButton';
import SearchBar from './components/SearchBar';
import { TSortOrder } from '../../types/home';
import SortFilterModal from './components/SortFilterModal';
import { ASC } from '../../utils/constants';

const HomeScreen = () => {
  const dispatch = useAppDispatch();
  const {
    repositories,
    searchRepositories,
    loading,
    loadingMore,
    searchLoading,
    error,
    hasMore,
  } = useAppSelector(state => state.repository);

  const [isSearching, setIsSearching] = useState(false);
  const [showSearchResults, setShowSearchResults] = useState(false);
  const [searchQuery, setSearchQuery] = useState('');
  const [sortOrder, setSortOrder] = useState<TSortOrder>(ASC);
  const [modalVisible, setModalVisible] = useState(false);

  useEffect(() => {
    dispatch(fetchRepositoriesRequest({ since: 0 }));
  }, [dispatch]);

  const sortedData = useMemo(() => {
    const dataToSort = showSearchResults ? searchRepositories : repositories;
    const sorted = [...dataToSort].sort((a, b) => {
      if (sortOrder === ASC) {
        return a.name.localeCompare(b.name);
      } else {
        return b.name.localeCompare(a.name);
      }
    });
    return sorted;
  }, [showSearchResults, searchRepositories, repositories, sortOrder]);

  const handleRefresh = () => {
    if (showSearchResults) {
      handleSubmitSearch(searchQuery);
    } else {
      dispatch(fetchRepositoriesRequest({ since: 0 }));
    }
  };

  const handleLoadMore = () => {
    if (!showSearchResults && !loadingMore && hasMore && !loading) {
      dispatch(fetchMoreRepositoriesRequest());
    }
  };

  const handleSearch = (query: string) => {
    setSearchQuery(query);
    if (query.trim().length > 0) {
      setIsSearching(true);
      dispatch(fetchSearchRepositoriesRequest({ query }));
    }
  };
  const handleSubmitSearch = (query: string) => {
    setSearchQuery(query);
    setShowSearchResults(true);
    setIsSearching(true);
    if (query.trim().length > 0) {
      dispatch(fetchSearchRepositoriesRequest({ query }));
    }
  };

  const handleSelectSuggestion = (repository: IRepository) => {
    setSearchQuery(repository.name);
    setShowSearchResults(true);
    setIsSearching(true);
    dispatch(fetchSearchRepositoriesRequest({ query: repository.name }));
  };

  const handleClearSearch = () => {
    setIsSearching(false);
    setShowSearchResults(false);
    setSearchQuery('');
    dispatch(clearSearchResults());
  };

  const handleOpenModal = () => {
    setModalVisible(true);
  };

  const handleCloseModal = () => {
    setModalVisible(false);
  };

  const handleSort = (order: TSortOrder) => {
    setSortOrder(order);
  };

  const renderItem = ({ item }: { item: IRepository }) => (
    <CardRepository item={item} />
  );

  const renderFooter = () => {
    if (!loadingMore) return null;
    return (
      <View style={styles.footerLoader}>
        <ActivityIndicator size="small" color={colors.shade800} />
      </View>
    );
  };

  const renderEmpty = () => {
    if (loading && !showSearchResults) {
      return (
        <View style={styles.emptyContainer}>
          <ActivityIndicator size="large" color={colors.shade800} />
          <Text style={styles.loadingText}>Loading repositories...</Text>
        </View>
      );
    }

    if (showSearchResults && searchLoading) {
      return (
        <View style={styles.emptyContainer}>
          <ActivityIndicator size="large" color={colors.shade800} />
          <Text style={styles.loadingText}>Searching...</Text>
        </View>
      );
    }

    if (
      showSearchResults &&
      searchRepositories.length === 0 &&
      !searchLoading
    ) {
      return (
        <View style={styles.emptyContainer}>
          <Text style={styles.noDataText}>No Data Found</Text>
        </View>
      );
    }

    if (error) {
      return (
        <View style={styles.emptyContainer}>
          <Text style={styles.errorText}>{error}</Text>
        </View>
      );
    }

    return null;
  };

  return (
    <SafeAreaView style={styles.container}>
      <SearchBar
        onSearch={handleSearch}
        onSelectSuggestion={handleSelectSuggestion}
        onClear={handleClearSearch}
        onSubmitSearch={handleSubmitSearch}
        suggestions={
          isSearching && !showSearchResults ? searchRepositories : []
        }
        loading={searchLoading}
        placeholder="Search GitHub repository..."
      />

      <FlatList
        data={sortedData}
        renderItem={renderItem}
        keyExtractor={item => item.id.toString()}
        ItemSeparatorComponent={ItemSeparator}
        showsVerticalScrollIndicator={false}
        style={styles.flatList}
        refreshControl={
          <RefreshControl
            refreshing={
              (loading && repositories.length === 0) ||
              (searchLoading && showSearchResults)
            }
            onRefresh={handleRefresh}
          />
        }
        onEndReached={handleLoadMore}
        onEndReachedThreshold={0.5}
        ListFooterComponent={renderFooter}
        ListEmptyComponent={renderEmpty}
      />

      <FloatingActionButton onPress={handleOpenModal} />

      <SortFilterModal
        visible={modalVisible}
        onClose={handleCloseModal}
        onSort={handleSort}
        currentOrder={sortOrder}
      />
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.brokenWhite,
  },
  header: {
    backgroundColor: colors.white,
    padding: 8,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  flatList: {
    paddingVertical: 8,
    paddingHorizontal: 8,
  },
  footerLoader: {
    paddingVertical: 20,
    alignItems: 'center',
  },
  emptyContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    paddingVertical: 50,
  },
  errorText: {
    fontSize: 16,
    color: colors.shade800,
    textAlign: 'center',
  },
  loadingText: {
    marginTop: 10,
    fontSize: 14,
    color: colors.shade75,
  },
  noDataText: {
    fontSize: 20,
    fontWeight: '600',
    color: colors.shade800,
    marginBottom: 8,
  },
  noDataSubtext: {
    fontSize: 14,
    color: colors.shade75,
    textAlign: 'center',
  },
});

export default HomeScreen;
