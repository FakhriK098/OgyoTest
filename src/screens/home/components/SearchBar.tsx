import React, { useState, useCallback, useEffect } from 'react';
import {
  View,
  TextInput,
  Text,
  StyleSheet,
  TouchableOpacity,
  FlatList,
  ActivityIndicator,
  Keyboard,
  TouchableWithoutFeedback,
} from 'react-native';
import { ISearchBarProps } from '../../../types/home';
import { IRepository } from '../../../types/repository';
import { colors } from '../../../themes/colors';

const SearchBar: React.FC<ISearchBarProps> = ({
  onSearch,
  onSelectSuggestion,
  onClear,
  suggestions,
  loading,
  placeholder = 'Search repository...',
  onSubmitSearch,
}) => {
  const [query, setQuery] = useState('');
  const [showSuggestions, setShowSuggestions] = useState(false);
  const [debounceTimer, setDebounceTimer] = useState<NodeJS.Timeout | null>(
    null,
  );

  useEffect(() => {
    return () => {
      if (debounceTimer) {
        clearTimeout(debounceTimer);
      }
    };
  }, [debounceTimer]);

  const handleTextChange = useCallback(
    (text: string) => {
      setQuery(text);

      if (debounceTimer) {
        clearTimeout(debounceTimer);
      }

      if (text.trim().length > 0) {
        setShowSuggestions(true);
        const timer = setTimeout(() => {
          onSearch(text.trim());
        }, 500);
        setDebounceTimer(timer);
      } else {
        setShowSuggestions(false);
        onClear();
      }
    },
    [debounceTimer, onSearch, onClear],
  );

  const handleSelectSuggestion = useCallback(
    (item: IRepository) => {
      setQuery(item.name);
      setShowSuggestions(false);
      onSelectSuggestion(item);
      Keyboard.dismiss();
    },
    [onSelectSuggestion],
  );

  const handleClear = useCallback(() => {
    setQuery('');
    setShowSuggestions(false);
    onClear();
  }, [onClear]);

  const handleSubmitEditing = useCallback(() => {
    if (query.trim().length > 0) {
      setShowSuggestions(false);
      onSubmitSearch(query.trim());
      Keyboard.dismiss();
    }
  }, [query, onSubmitSearch]);

  const renderSuggestion = ({ item }: { item: IRepository }) => (
    <TouchableOpacity
      style={styles.suggestionItem}
      onPress={() => handleSelectSuggestion(item)}
    >
      <Text style={styles.suggestionName} numberOfLines={1}>
        {item.name}
      </Text>
    </TouchableOpacity>
  );

  const renderSuggestionSeparator = () => (
    <View style={styles.suggestionSeparator} />
  );

  const renderEmptySuggestions = () => {
    if (loading) {
      return (
        <View style={styles.loadingContainer}>
          <ActivityIndicator size="small" color={colors.shade800} />
          <Text style={styles.loadingText}>Loading suggestions...</Text>
        </View>
      );
    }

    if (query.trim().length > 0 && !loading && suggestions.length === 0) {
      return (
        <View style={styles.noDataContainer}>
          <Text style={styles.noDataText}>No suggestions found</Text>
        </View>
      );
    }

    return null;
  };

  return (
    <>
      <View style={styles.container}>
        <View style={styles.searchContainer}>
          <View style={styles.searchInputContainer}>
            <TextInput
              style={styles.searchInput}
              placeholder={placeholder}
              value={query}
              onChangeText={handleTextChange}
              onFocus={() =>
                query.trim().length > 0 && setShowSuggestions(true)
              }
              autoCapitalize="none"
              autoCorrect={false}
              returnKeyType="search"
              onSubmitEditing={handleSubmitEditing}
            />
            {query.length > 0 && (
              <TouchableOpacity
                onPress={handleClear}
                style={styles.clearButton}
              >
                <Text style={styles.clearIcon}>âœ•</Text>
              </TouchableOpacity>
            )}
          </View>
        </View>

        {showSuggestions && query.trim().length > 0 && (
          <TouchableWithoutFeedback onPress={() => setShowSuggestions(false)}>
            <View style={styles.suggestionsOverlay}>
              <TouchableWithoutFeedback>
                <View style={styles.suggestionsContainer}>
                  <FlatList
                    data={suggestions.slice(0, 8)}
                    renderItem={renderSuggestion}
                    keyExtractor={item => item.id.toString()}
                    ItemSeparatorComponent={renderSuggestionSeparator}
                    ListEmptyComponent={renderEmptySuggestions}
                    keyboardShouldPersistTaps="always"
                    style={styles.suggestionsList}
                    contentContainerStyle={styles.suggestionsContent}
                    scrollEnabled={suggestions.length > 5}
                  />
                </View>
              </TouchableWithoutFeedback>
            </View>
          </TouchableWithoutFeedback>
        )}
      </View>
    </>
  );
};

const styles = StyleSheet.create({
  container: {
    backgroundColor: colors.white,
    zIndex: 1000,
  },
  searchContainer: {
    paddingHorizontal: 8,
    paddingVertical: 12,
    borderBottomWidth: 1,
    borderBottomColor: colors.shade75,
  },
  searchInputContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    borderWidth: 1,
    borderColor: colors.shade800,
    borderRadius: 8,
    paddingHorizontal: 12,
    height: 40,
  },
  searchInput: {
    flex: 1,
    fontSize: 16,
    color: colors.black,
    paddingVertical: 0,
  },
  clearButton: {
    padding: 4,
  },
  clearIcon: {
    fontSize: 18,
    color: colors.shade800,
  },
  suggestionsOverlay: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: -1000,
    zIndex: 1001,
  },
  suggestionsContainer: {
    position: 'absolute',
    top: 64,
    left: 8,
    right: 8,
    backgroundColor: colors.white,
    borderRadius: 8,
    maxHeight: 280,
    shadowColor: colors.black,
    shadowOffset: {
      width: 0,
      height: 4,
    },
    shadowOpacity: 0.3,
    shadowRadius: 4.65,
    elevation: 8,
  },
  suggestionsList: {
    backgroundColor: 'transparent',
    borderRadius: 8,
  },
  suggestionsContent: {
    paddingVertical: 4,
  },
  suggestionItem: {
    paddingHorizontal: 16,
    paddingVertical: 14,
    backgroundColor: colors.white,
  },
  suggestionName: {
    fontSize: 15,
    color: colors.black,
    fontWeight: '500',
  },
  suggestionFullName: {
    fontSize: 13,
    color: colors.shade800,
    marginBottom: 2,
  },
  suggestionDescription: {
    fontSize: 12,
    color: colors.shade75,
    lineHeight: 16,
  },
  suggestionSeparator: {
    height: 1,
    backgroundColor: colors.brokenWhite,
    marginHorizontal: 16,
  },
  loadingContainer: {
    padding: 20,
    alignItems: 'center',
  },
  loadingText: {
    marginTop: 8,
    fontSize: 14,
    color: colors.shade75,
  },
  noDataContainer: {
    padding: 20,
    alignItems: 'center',
  },
  noDataText: {
    fontSize: 14,
    fontWeight: '500',
    color: colors.shade800,
    marginBottom: 4,
  },
  noDataSubtext: {
    fontSize: 12,
    color: colors.shade75,
  },
});

export default SearchBar;
