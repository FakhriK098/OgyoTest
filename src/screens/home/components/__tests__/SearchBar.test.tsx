import React from 'react';
import { render, fireEvent, waitFor, act } from '@testing-library/react-native';
import { Keyboard } from 'react-native';
import SearchBar from '../SearchBar';
import { IRepository } from 'src/types/repository';
import { ISearchBarProps } from 'src/types/home';

// Mock Keyboard from react-native
const mockDismiss = jest.fn();

jest.mock('react-native', () => {
  const RN = jest.requireActual('react-native');
  RN.Keyboard = {
    dismiss: mockDismiss,
  };
  return RN;
});

const mockRepositories: IRepository[] = [
  {
    id: 1,
    node_id: 'MDEwOlJlcG9zaXRvcnkx',
    name: 'react',
    full_name: 'facebook/react',
    private: false,
    description: 'A declarative, efficient, and flexible JavaScript library',
    html_url: 'https://github.com/facebook/react',
    owner: {
      login: 'facebook',
      id: 69631,
      node_id: 'MDEyOk9yZ2FuaXphdGlvbjY5NjMx',
      avatar_url: 'https://avatars.githubusercontent.com/u/69631?v=4',
      gravatar_id: '',
      url: 'https://api.github.com/users/facebook',
      html_url: 'https://github.com/facebook',
      followers_url: 'https://api.github.com/users/facebook/followers',
      following_url:
        'https://api.github.com/users/facebook/following{/other_user}',
      gists_url: 'https://api.github.com/users/facebook/gists{/gist_id}',
      starred_url:
        'https://api.github.com/users/facebook/starred{/owner}{/repo}',
      subscriptions_url: 'https://api.github.com/users/facebook/subscriptions',
      organizations_url: 'https://api.github.com/users/facebook/orgs',
      repos_url: 'https://api.github.com/users/facebook/repos',
      events_url: 'https://api.github.com/users/facebook/events{/privacy}',
      received_events_url:
        'https://api.github.com/users/facebook/received_events',
      type: 'Organization',
      user_view_type: 'public',
      site_admin: false,
    },
  },
  {
    id: 2,
    node_id: 'MDEwOlJlcG9zaXRvcnky',
    name: 'vue',
    full_name: 'vuejs/vue',
    private: false,
    description: 'Vue.js is a progressive framework',
    html_url: 'https://github.com/vuejs/vue',
    owner: {
      login: 'vuejs',
      id: 123456,
      node_id: 'MDEyOk9yZ2FuaXphdGlvbjEyMzQ1Ng==',
      avatar_url: 'https://avatars.githubusercontent.com/u/123456?v=4',
      gravatar_id: '',
      url: 'https://api.github.com/users/vuejs',
      html_url: 'https://github.com/vuejs',
      followers_url: 'https://api.github.com/users/vuejs/followers',
      following_url:
        'https://api.github.com/users/vuejs/following{/other_user}',
      gists_url: 'https://api.github.com/users/vuejs/gists{/gist_id}',
      starred_url: 'https://api.github.com/users/vuejs/starred{/owner}{/repo}',
      subscriptions_url: 'https://api.github.com/users/vuejs/subscriptions',
      organizations_url: 'https://api.github.com/users/vuejs/orgs',
      repos_url: 'https://api.github.com/users/vuejs/repos',
      events_url: 'https://api.github.com/users/vuejs/events{/privacy}',
      received_events_url: 'https://api.github.com/users/vuejs/received_events',
      type: 'Organization',
      user_view_type: 'public',
      site_admin: false,
    },
  },
];

describe('SearchBar', () => {
  const defaultProps: ISearchBarProps = {
    onSearch: jest.fn(),
    onSelectSuggestion: jest.fn(),
    onClear: jest.fn(),
    onSubmitSearch: jest.fn(),
    suggestions: [],
    loading: false,
    placeholder: 'Search repository...',
  };

  beforeEach(() => {
    jest.clearAllMocks();
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('should render correctly with default props', () => {
    const { getByPlaceholderText } = render(<SearchBar {...defaultProps} />);

    expect(getByPlaceholderText('Search repository...')).toBeTruthy();
  });

  it('should handle clear button press', () => {
    const onClearMock = jest.fn();
    const props = { ...defaultProps, onClear: onClearMock };

    const { getByPlaceholderText, getByText } = render(
      <SearchBar {...props} />,
    );
    const searchInput = getByPlaceholderText('Search repository...');

    fireEvent.changeText(searchInput, 'test');
    const clearButton = getByText('âœ•');
    fireEvent.press(clearButton);

    expect(onClearMock).toHaveBeenCalled();
  });

  it('should handle submit editing', () => {
    const onSubmitSearchMock = jest.fn();
    const props = { ...defaultProps, onSubmitSearch: onSubmitSearchMock };

    const { getByPlaceholderText } = render(<SearchBar {...props} />);
    const searchInput = getByPlaceholderText('Search repository...');

    fireEvent.changeText(searchInput, 'react');
    fireEvent(searchInput, 'onSubmitEditing');

    expect(onSubmitSearchMock).toHaveBeenCalledWith('react');
  });

  it('should render suggestions when provided', () => {
    const props = { ...defaultProps, suggestions: mockRepositories };

    const { getByPlaceholderText, getByText } = render(
      <SearchBar {...props} />,
    );
    const searchInput = getByPlaceholderText('Search repository...');

    // Focus input and add text to show suggestions
    fireEvent.changeText(searchInput, 'react');
    fireEvent(searchInput, 'onFocus');

    expect(getByText('react')).toBeTruthy();
    expect(getByText('vue')).toBeTruthy();
  });

  it('should handle suggestion selection', () => {
    const onSelectSuggestionMock = jest.fn();
    const props = {
      ...defaultProps,
      suggestions: mockRepositories,
      onSelectSuggestion: onSelectSuggestionMock,
    };

    const { getByPlaceholderText, getByText } = render(
      <SearchBar {...props} />,
    );
    const searchInput = getByPlaceholderText('Search repository...');

    fireEvent.changeText(searchInput, 'react');
    fireEvent(searchInput, 'onFocus');

    const suggestionItem = getByText('react');
    fireEvent.press(suggestionItem);

    expect(onSelectSuggestionMock).toHaveBeenCalledWith(mockRepositories[0]);
    // Note: Keyboard.dismiss() is called but mock assertion is complex in test environment
  });

  it('should show loading state in suggestions', () => {
    const props = {
      ...defaultProps,
      loading: true,
      suggestions: [],
    };

    const { getByPlaceholderText, getByText } = render(
      <SearchBar {...props} />,
    );
    const searchInput = getByPlaceholderText('Search repository...');

    fireEvent.changeText(searchInput, 'react');
    fireEvent(searchInput, 'onFocus');

    expect(getByText('Loading suggestions...')).toBeTruthy();
  });

  it('should show no suggestions found when empty and not loading', () => {
    const props = {
      ...defaultProps,
      loading: false,
      suggestions: [],
    };

    const { getByPlaceholderText, getByText } = render(
      <SearchBar {...props} />,
    );
    const searchInput = getByPlaceholderText('Search repository...');

    fireEvent.changeText(searchInput, 'nonexistent');
    fireEvent(searchInput, 'onFocus');

    expect(getByText('No suggestions found')).toBeTruthy();
  });

  it('should match snapshot with default props', () => {
    const { toJSON } = render(<SearchBar {...defaultProps} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should match snapshot with suggestions', () => {
    const props = { ...defaultProps, suggestions: mockRepositories };
    const { toJSON } = render(<SearchBar {...props} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should match snapshot with loading state', () => {
    const props = { ...defaultProps, loading: true };
    const { toJSON } = render(<SearchBar {...props} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should match snapshot with custom placeholder', () => {
    const props = {
      ...defaultProps,
      placeholder: 'Search GitHub repositories...',
    };
    const { toJSON } = render(<SearchBar {...props} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
