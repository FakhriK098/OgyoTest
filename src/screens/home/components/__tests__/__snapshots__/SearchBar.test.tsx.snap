// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should match snapshot with custom placeholder 1`] = `
<View
  style={
    {
      "backgroundColor": "#FFFFFF",
      "zIndex": 1000,
    }
  }
>
  <View
    style={
      {
        "borderBottomColor": "#DEDEDE",
        "borderBottomWidth": 1,
        "paddingHorizontal": 8,
        "paddingVertical": 12,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "borderColor": "#161B28",
          "borderRadius": 8,
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 40,
          "paddingHorizontal": 12,
        }
      }
    >
      <TextInput
        autoCapitalize="none"
        autoCorrect={false}
        onChangeText={[Function]}
        onFocus={[Function]}
        onSubmitEditing={[Function]}
        placeholder="Search GitHub repositories..."
        returnKeyType="search"
        style={
          {
            "color": "#000000",
            "flex": 1,
            "fontSize": 16,
            "paddingVertical": 0,
          }
        }
        value=""
      />
    </View>
  </View>
</View>
`;

exports[`SearchBar should match snapshot with default props 1`] = `
<View
  style={
    {
      "backgroundColor": "#FFFFFF",
      "zIndex": 1000,
    }
  }
>
  <View
    style={
      {
        "borderBottomColor": "#DEDEDE",
        "borderBottomWidth": 1,
        "paddingHorizontal": 8,
        "paddingVertical": 12,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "borderColor": "#161B28",
          "borderRadius": 8,
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 40,
          "paddingHorizontal": 12,
        }
      }
    >
      <TextInput
        autoCapitalize="none"
        autoCorrect={false}
        onChangeText={[Function]}
        onFocus={[Function]}
        onSubmitEditing={[Function]}
        placeholder="Search repository..."
        returnKeyType="search"
        style={
          {
            "color": "#000000",
            "flex": 1,
            "fontSize": 16,
            "paddingVertical": 0,
          }
        }
        value=""
      />
    </View>
  </View>
</View>
`;

exports[`SearchBar should match snapshot with loading state 1`] = `
<View
  style={
    {
      "backgroundColor": "#FFFFFF",
      "zIndex": 1000,
    }
  }
>
  <View
    style={
      {
        "borderBottomColor": "#DEDEDE",
        "borderBottomWidth": 1,
        "paddingHorizontal": 8,
        "paddingVertical": 12,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "borderColor": "#161B28",
          "borderRadius": 8,
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 40,
          "paddingHorizontal": 12,
        }
      }
    >
      <TextInput
        autoCapitalize="none"
        autoCorrect={false}
        onChangeText={[Function]}
        onFocus={[Function]}
        onSubmitEditing={[Function]}
        placeholder="Search repository..."
        returnKeyType="search"
        style={
          {
            "color": "#000000",
            "flex": 1,
            "fontSize": 16,
            "paddingVertical": 0,
          }
        }
        value=""
      />
    </View>
  </View>
</View>
`;

exports[`SearchBar should match snapshot with suggestions 1`] = `
<View
  style={
    {
      "backgroundColor": "#FFFFFF",
      "zIndex": 1000,
    }
  }
>
  <View
    style={
      {
        "borderBottomColor": "#DEDEDE",
        "borderBottomWidth": 1,
        "paddingHorizontal": 8,
        "paddingVertical": 12,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "borderColor": "#161B28",
          "borderRadius": 8,
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 40,
          "paddingHorizontal": 12,
        }
      }
    >
      <TextInput
        autoCapitalize="none"
        autoCorrect={false}
        onChangeText={[Function]}
        onFocus={[Function]}
        onSubmitEditing={[Function]}
        placeholder="Search repository..."
        returnKeyType="search"
        style={
          {
            "color": "#000000",
            "flex": 1,
            "fontSize": 16,
            "paddingVertical": 0,
          }
        }
        value=""
      />
    </View>
  </View>
</View>
`;
